### DISCUSSION TITLE
Quality codes

### DISCUSSION THREAD

**Discussion by jasperlg** [thomassorensen2650/node-red-contrib-mqtt-sparkplug-plus#31] [Q&A]
Hi @thomassorensen2650,

I'm wondering, I don't see anything in the code for quality codes. Does that mean passing quality codes will not work?  Or should they all work? I've done some tests, but I don't see it working at the moment.

**Comment by thomassorensen2650** [comment]
Hi

This is not implemented. normally you'd just send NULL on bad data. Is there a usecase where it adds value to have more control over quality codes?

**Reply by jasperlg** [reply]
Yes, it does add value. if an application (like Ignition f.e.) is listening to the sparkplug messages it has build in functionality to handle quality codes from sparkplug. 

In my use case I connect from Node-Red to Modbus devices. If these devices do not respond, I want to send the bad quality code over to my broker so all applications know not to trust the data anymore.

Plus, it's also in the spec. 
[tck-id-payloads-propertyset-quality-value-type] The type of the Property Value MUST be a value of 3 which represents a Signed 32-bit Integer. • [tck-id-payloads-propertyset-quality-value-value] The value of the Property Value MUST be an int_value and be one of the valid quality codes of 0, 192, or 500.

**Comment by thomassorensen2650** [comment]
I think the correct way to handle that situation would be to send a DDEATH message if the modbus device is not connected.  

But QualityCodes are part of the specification, so it should work. I just created a quick local test with Ignition, and it looks like qualify codes works as expected (see example below). 

```
{
    "metrics": [
        {
            "name": "testing/test1",
            "value": 12,
            "properties": {
                "Quality": {
                    "type": "Int32",
                    "value": -2147483136
                }
            }
        }
    ]
}
```

With never version of Igntition you might need to send a value per the SPB v3.0 spec. (e.g. 500)
